File: backend/app/api/stats.py
"""统计接口（M3+M7，DR-020）：趋势时间序列 + 词云/作者榜/热力图/合作网络。"""
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.db import get_db
from app.schemas.stats import (
    AuthorPage,
    InstitutionDetailResponse,
    InstitutionPage,
    TrendResponse,
)
from app.services import institution_service, stats_service

router = APIRouter(prefix="/stats", tags=["stats"])


@router.get("/trends", response_model=TrendResponse)
def trends(
    group_by: str = Query("topic", pattern="^(topic|venue|year)$", description="分组维度"),
    start: str | None = Query(None, description="起始日期 YYYY-MM-DD（默认数据最早日）"),
    end: str | None = Query(None, description="结束日期 YYYY-MM-DD（默认数据最晚日）"),
    db: Session = Depends(get_db),
) -> dict:
    return stats_service.trends(db, group_by, start, end)


@router.get("/words")
def words(
    limit: int = Query(50, ge=10, le=200, description="词云词数上限"),
    scope: str = Query("ai4se", pattern="^(all|ai4se)$", description="统计范围"),
    db: Session = Depends(get_db),
) -> dict:
    return stats_service.words(db, limit, scope)


@router.get("/authors", response_model=AuthorPage)
def authors_top(
    page: int = Query(1, ge=1, description="页码（从 1 起）"),
    page_size: int = Query(20, ge=1, le=100, description="每页条数"),
    q: str | None = Query(None, description="按作者名模糊搜索（不区分大小写）"),
    db: Session = Depends(get_db),
) -> dict:
    items, total = stats_service.authors_top(db, page, page_size, q)
    return {"items": items, "total": total, "page": page, "page_size": page_size}


@router.get("/institutions", response_model=InstitutionPage)
def institutions_top(
    page: int = Query(1, ge=1, description="页码（从 1 起）"),
    page_size: int = Query(20, ge=1, le=100, description="每页条数"),
    q: str | None = Query(None, description="按机构名模糊搜索（不区分大小写）"),
    db: Session = Depends(get_db),
) -> dict:
    items, total = stats_service.institutions_top(db, page, page_size, q)
    return {"items": items, "total": total, "page": page, "page_size": page_size}


@router.get("/institution", response_model=InstitutionDetailResponse)
def institution_detail(
    name: str = Query(..., min_length=1, description="机构名（归一化后的完整机构串）"),
    db: Session = Depends(get_db),
) -> dict:
    """机构详情（M12）：统计 + 主题分布 + 合作机构。未知名返回零值（不 404）。"""
    return institution_service.institution_detail(db, name)


@router.get("/cross")
def cross(db: Session = Depends(get_db)) -> dict:
    """会议×主题交叉矩阵（热力图）。"""
    return stats_service.cross(db)


@router.get("/coauthor")
def coauthor(
    limit: int = Query(100, ge=20, le=300, description="合作网络节点（活跃作者）上限"),
    db: Session = Depends(get_db),
) -> dict:
    return stats_service.coauthor(db, limit)

File: frontend/src/api/stats.ts
import http from './http'
import type {
  AuthorPage, CoauthorGraph, CrossMatrix, InstitutionDetail, InstitutionPage,
  InstitutionStat, TrendResponse, WordItem,
} from '../types'

// 趋势：group_by=topic|venue|year；后端按天返回原始计数，聚合粒度由前端决定（DR-020）
export async function getTrends(
  groupBy: 'topic' | 'venue' | 'year',
  start?: string,
  end?: string,
): Promise<TrendResponse> {
  const { data } = await http.get<TrendResponse>('/stats/trends', {
    params: { group_by: groupBy, start, end },
  })
  return data
}

// ---- M7 分析图表 ----

export async function getWords(limit = 50, scope: 'all' | 'ai4se' = 'ai4se'): Promise<WordItem[]> {
  const { data } = await http.get<{ words: WordItem[] }>('/stats/words', {
    params: { limit, scope },
  })
  return data.words
}

// M13 作者/机构榜：服务端分页（返回 {items, total, page, page_size}）；q 按作者名模糊搜索
export async function getAuthorsTop(
  page = 1,
  pageSize = 20,
  q?: string,
): Promise<AuthorPage> {
  const { data } = await http.get<AuthorPage>('/stats/authors', {
    params: { page, page_size: pageSize, q: q || undefined },
  })
  return data
}

export async function getInstitutionsTop(
  page = 1,
  pageSize = 20,
  q?: string,
): Promise<InstitutionPage> {
  const { data } = await http.get<InstitutionPage>('/stats/institutions', {
    params: { page, page_size: pageSize, q: q || undefined },
  })
  return data
}

// PaperListView 机构自动补全数据源：分页循环取全部机构（后端 page_size 上限 100）
export async function getAllInstitutions(): Promise<InstitutionStat[]> {
  const first = await getInstitutionsTop(1, 100)
  const pages = Math.max(1, Math.ceil(first.total / 100))
  const rest = await Promise.all(
    Array.from({ length: pages - 1 }, (_, i) => getInstitutionsTop(i + 2, 100)),
  )
  return [...first.items, ...rest.flatMap((p) => p.items)]
}

// M12 机构详情：统计 + 主题分布 + 合作机构（axios 自动编码机构名）
export async function getInstitutionDetail(name: string): Promise<InstitutionDetail> {
  const { data } = await http.get<InstitutionDetail>('/stats/institution', { params: { name } })
  return data
}

export async function getCross(): Promise<CrossMatrix> {
  const { data } = await http.get<CrossMatrix>('/stats/cross')
  return data
}

export async function getCoauthor(limit = 100): Promise<CoauthorGraph> {
  const { data } = await http.get<CoauthorGraph>('/stats/coauthor', { params: { limit } })
  return data
}
